pkg/runtime/unstructured/converter.go

```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
 func (c *Converter) interfaceFromUnstructured(sv, dv reflect.Value) error {
 	return fmt.Errorf("Interface conversion unsupported")
 }
+
+var (
+	marshalerType          = reflect.TypeOf(new(encodingjson.Marshaler)).Elem()
+	mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+)
+
+func (c *Converter) ToUnstructured(obj runtime.Object, u *map[string]interface{}) error {
+	return c.toUnstructured(reflect.ValueOf(obj).Elem(), reflect.ValueOf(u).Elem())
+}
+
+func (c *Converter) toUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+
+	// Check if the object has a custom JSON marshaller/unmarshaller.
+	if st.Implements(marshalerType) {
+		var data []byte
+		var err error
+		if sv.Kind() == reflect.Ptr && sv.IsNil() {
+			data = []byte("null")
+		} else {
+			marshaler := sv.Interface().(encodingjson.Marshaler)
+			data, err = marshaler.MarshalJSON()
+			if err != nil {
+				return err
+			}
+		}
+		if string(data) == "null" {
```

Comment:
I think it would be better to compare `bytes.Equal(data, []byte("null"))` but the benchmark would tell us.